### PR TITLE
[INTENSE SUFFERING] [PORT] TGUI Techfabs, but Without The Lost Features!

### PR DESCRIPTION
**Pull request by RimiNosha**
## About The Pull Request
https://github.com/tgstation/tgstation/pull/69990
https://github.com/tgstation/tgstation/pull/72276
https://github.com/tgstation/tgstation/pull/72721

Yup. Reagents are listed now. It looks alright enough, though recipes with more than 3 reagents might look weird.

Also creeped in the order console sec > service PR before anyone can try to argue otherwise.

## How Does This Help ***Gameplay***?

Self evident. Nicer UI makes for nicer gameplay.

## Proof of Testing

<details>
<summary>Screenshots/Videos</summary> 

Recipe with required reagent
image

Recipe with missing reagent
image

Recipe with low reagents
image

MULTI REAGENT DRIFTING
image

</details>

## Changelog

:cl: scriptis (original), RimiNosha (port and reagents functionality)
refactor: mechfabs, autolathes, component printers, and module duplicators now use techfab tgui components
refactor: every single design is now categorized and subcategorized
refactor: mechfabs and autolathes are now in typescript
qol: techfabs now have icons for what you're about to print
qol: techfab material icons are now animated
qol: techfab material icons now fade when no materials are available
qol: techfab searching no longer lags like hell
qol: techfab searching now searches all recipes instead of just the current category
qol: techfabs now have subcategorization (stock part users rejoice)
qol: techfabs now announce when new recipes are available
qol: numerous other techfab ui tweaks
balance: some designs that were formerly autolathe exclusive can now be printed at some departmental techfabs
fix: Reagents display on the mechfab UI once again.
/:cl:

### CODE REVIEW

**Comment by nevimer** [comment]
in the process of reviewing

File: code/modules/wiremod/core/component_printer.dm [hunk 1]
```diff
@@ -422,12 +433,15 @@
...
 		designs["[index]"] = list(
 			"name" = design["name"],
-			"description" = design["desc"],
-			"materials" = get_material_cost_data(design["materials"]),
-			"categories" = list("Circuitry"),
+			"desc" = design["desc"],
+			"cost" = get_material_cost_data(design["materials"]),
+			"id" = "[index]",
+			"icon" = "integrated_circuit",
+			"categories" = list("/Saved Circuits"),
```
Comment (nevimer) [inline code/modules/wiremod/core/component_printer.dm:444]: / is intentional?

File: code/game/machinery/autolathe.dm [hunk 2]
```diff
@@ -59,171 +35,195 @@
...
+		if(disabled || !can_build(design))
 			unbuildable = TRUE
+
 		var/max_multiplier = unbuildable ? 0 : 1
-		if(ispath(D.build_path, /obj/item/stack))
+
+		if(ispath(design.build_path, /obj/item/stack))
 			sheets = TRUE
+
 			if(!unbuildable)
```
Comment (nevimer) [inline code/game/machinery/autolathe.dm:91]: if this is possibly set by the power being out, what could this imply?

File: code/game/machinery/autolathe.dm [hunk 3]
```diff
@@ -59,171 +35,195 @@
...
-			else
-				to_chat(usr, span_alert("Not enough materials for this operation."))
-		else
+		var/datum/design/design = SSresearch.techweb_design_by_id(design_id)
+
+		if(!(design.build_type & AUTOLATHE) || design.id != design_id)
+			return
+
+		if (busy)
 			to_chat(usr, span_alert("The autolathe is busy. Please wait for completion of previous operation."))
```
Comment (nevimer) [inline code/game/machinery/autolathe.dm:163]: does it use this? line 233 seems to be very similar

File: code/modules/research/designs/medical_designs.dm [hunk 4]
```diff
@@ -8,29 +8,9 @@
 	build_type = PROTOLATHE | AWAY_LATHE
 	materials = list(/datum/material/iron = 500, /datum/material/glass = 50)
 	build_path = /obj/item/healthanalyzer
-	category = list(RND_CATEGORY_MEDICAL_DESIGNS)
-	departmental_flags = DEPARTMENT_BITFLAG_MEDICAL
-
-/datum/design/mmi
```
[the remark this answers is not in the document]
Comment (RimiNosha) [reply code/modules/research/designs/medical_designs.dm:14]: Yup. To be fair, medical starts with a handful, and they're almost never used in medical aside from validing.

File: code/modules/wiremod/core/component_printer.dm (hunk 1 above)
Comment (RimiNosha) [reply code/modules/wiremod/core/component_printer.dm:444]: Yes, this is a TG thing, the `/` should be treated like you're defining an absolute path.

File: code/game/machinery/autolathe.dm (hunk 2 above)
Comment (RimiNosha) [reply code/game/machinery/autolathe.dm:91]: Literally decides if a design should be built after all checks have ran. Also 1:1 with TG lathe logic.

File: code/game/machinery/autolathe.dm (hunk 3 above)
Comment (RimiNosha) [reply code/game/machinery/autolathe.dm:163]: Yes, it does. It's to stop spamming the lathe, cause creating new in-world atoms is surprisingly expensive, especially when you have to move them. See the many, **many** issues TG and SR have with empty bottles flooding the halls and murdering the server for an example of why this is needed.

**Review by nevimer — APPROVED** [review]

**Pull request merged**